### ISSUE TITLE
Excessive dependencies of prometheus_client

### ISSUE THREAD

**Issue by rolweber** [ContinuumIO/anaconda-issues#9789]
I'm upgrading Jupyter `notebook` from 5.4.1 to the latest version 5.6.0. There's a new dependency on `prometheus_client`. When I fetch the packages with `pip` from PyPI, I get just the new package. When I fetch the packages with `conda`, I also get `twisted` and about a dozen transitive dependencies.

The [prometheus_client `setup.py`](https://github.com/prometheus/client_python/blob/v0.3.0/setup.py) declares the dependency on `twisted` as an extra:
```python
    extras_require={
        'twisted': ['twisted'],
    },
```
whereas the conda package seems to have a hard dependency.

The conda package is also a version behind, at 0.2.0, while PyPI already has 0.3.0. There was no change in the dependencies between these versions.

Would it be possible to upgrade `prometheus_client` to 0.3.0 and remove the hard dependency on `twisted` for the new version?

**Comment by mingwandroid** [comment]
Twisted is not an 'excessive dependency', it's an 'extra requirement', please see https://github.com/prometheus/client_python/blob/master/setup.py#L20-L22

**Comment by rolweber** [comment]
So maybe I'm misunderstanding something... if I install the packages with `pip` instead of `conda`, is the installation broken because of the missing 'extra requirement'?

I generally prefer `conda` over `pip`, because I think you're doing a great job with your dependency management. But if that means pulling in a dozen extra packages into the Docker images I have to build, I have to wonder.

**Comment by rolweber** [comment]
As I understand this documentation 'extras' are optional. The conda package as it is being built has the dependency as mandatory. That's what I would call excessive, because I need to install a dozen packages which I don't actually need.

http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

**Comment by mingwandroid** [comment]
Extras are optional, but we prefer to make our software fully-featured.

**Comment by mingwandroid** [comment]
We cannot change such things without causing a lot of disruption for our users, not that I would want to anyway.

Asking the rest of the team for input on this though: @nehaljwani, @jjhelmus, @msarahan, @kalefranz

**Comment by rolweber** [comment]
OK, thanks for the explanation. I would really prefer to install Jupyter Notebook without Twisted, as the Docker images are for a hosted solution where the latter wouldn't work anyway. But I can do that from PyPI if necessary.

**Comment by rolweber** [comment]
I've given it some more thought. From my perspective, the situation is similar to Tensorflow/Tensorboard. A user running Tensorflow locally most likely needs Tensorboard as well. So there's a package `tensorflow` that installs both. But there's also a package `tensorflow-base`, for those who need only Tensorflow.
I had a problem with `keras` depending on `tensorflow`. This was solved in #9483 by introducing `keras-base`, which does not pull in optional dependencies. So anyone installing `keras` still gets it along with Tensorflow and Tensorboard, but those who don't want Tensorboard can install `keras-base` and `tensorflow-base`.

Anaconda already has a metapackage `jupyter`, which would be a good place to pull in `twisted`. Somebody who runs
```bash
conda install jupyter
```
wants the full user experience, and should get it. But I am deliberately avoiding that metapackage. I need a notebook server, that's the only port we'll expose anyway. So I run
```bash
conda install notebook
```
hoping to get the notebook server and everything I actually need to run it, but nothing more.

In the Jupyter ecosystem, although not in conda mainland, there's also `jupyter_kernel_gateway`. I know it's not the most vibrant project, but it exists. And it [depends on `notebook`](https://github.com/conda-forge/jupyter_kernel_gateway-feedstock/blob/master/recipe/meta.yaml#L28). Which means it will now pull in `twisted`, although the kernel gateway is meant for server-side scenarios where users (rather than developers) are not in a position to start extra server processes. The enterprise gateway might also be affected, I didn't check.

IMHO, dragging `twisted` as a transitive dependency of `notebook` into the Jupyter dependency graph, rather than making it an explicit dependency on a higher level that can be avoided, is the wrong approach. But I can work around it through pip, so I'm not going to pursue this.

**Comment by jjhelmus** [comment]
I've raised the question of dropping the `twisted` requirement in `prometheus_client` in the conda-forge feedstock, conda-forge/prometheus_client-feedstock#11.  Let us see what the maintainers there suggest.
